File: wrappers/android/src/main/java/net/devolutions/slauth/Totp.java
package net.devolutions.slauth;

import java.io.IOException;

public class Totp extends RustObject {
    static {
        System.loadLibrary("slauth");
    }

    public Totp(String uri) throws Exception {
        this.raw = JNA.INSTANCE.totp_from_uri(uri);
        if (this.raw == null) {
            throw new Exception();
        }
    }

    public String gen() {
        return JNA.INSTANCE.totp_gen(raw);
    }

    public String genWith(long elapsed) {
        return JNA.INSTANCE.totp_gen_with(raw, elapsed);
    }

    public String toUri(String label, String issuer) {
        return JNA.INSTANCE.totp_to_uri(raw, label, issuer);
    }

    public Boolean validateCurrent(String code) {
        return JNA.INSTANCE.totp_validate_current(raw, code);
    }

    public Boolean verify(String code) {
        return JNA.INSTANCE.totp_verify(raw, code);
    }

    @Override
    public void close() throws IOException {
        JNA.INSTANCE.totp_free(raw);
    }
}

File: wrappers/swift/classes/Totp.swift
//
//  Totp.swift
//  firebase
//
//  Created by Richer Archambault on 2019-04-26.
//  Copyright © 2019 Sebastien Aubin. All rights reserved.
//

import Foundation

public class Totp: NSObject, RustObject {
	var raw: OpaquePointer
	
	required init(raw: OpaquePointer) {
		self.raw = raw
	}
	
	func intoRaw() -> OpaquePointer {
		return self.raw
	}
	
	public convenience init(uri: String) throws {
		let r = totp_from_uri(uri)
        if r == nil {
            throw Err(message: "InvalidUri")
        } else {
            self.init(raw: r!)
        }
	}
	
	deinit {
		totp_free(raw)
	}
	
	public func to_uri(label: String, issuer: String) -> String {
		let uri = totp_to_uri(raw, label, issuer)
		let s = String(cString: uri!)
		free(uri)
		return s
	}
	
	public func gen() -> String {
		let code = totp_gen(raw)
		let s = String(cString: code!)
		free(code)
		return s
	}
	
	public func gen_with(elapsed: UInt) -> String {
		let code = totp_gen_with(raw, elapsed)
		let s = String(cString: code!)
		free(code)
		return s
	}
	
	public func verify(code: String) -> Bool {
		return totp_verify(raw, code)
	}
	
	public func validate_current(code: String) -> Bool {
		return totp_validate_current(raw, code)
	}
}
